### PR TITLE
Stats: Introducing redirect flow to all Stats pages

### PR DESCRIPTION
**Pull request by grzegorz-cp**
Related to #86976 #86929

## Proposed Changes

* Introduce a new component capturing access requirements added to the main Traffic page
* Wrap all pages in the `StatsRedirectFlow` component to avoid page flicking before redirecting

## Testing Instructions

* create a JN and connect it
* navigate to the live branch and apply `stats/checkout-flows-v2` feature flag
* navigate to the stats page and verify that you are redirected to the purchase page
* repeat the process by going to different stats pages like `Insights`, `Subscribers` or summary pages
* click the `I will do it later` button from the purchase page (note, there will be a delay that will use a loader in the future)
* verify that you are redirected to the Traffic page
* verify that you are not redirected to the purchase page for your old blogs

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/stats/stats-reditect-flow/index.tsx [hunk 1]
```diff
@@ -74,8 +77,12 @@ const StatsRedirectFlow = () => {
 
 		page.redirect( `/stats/purchase/${ siteSlug }?${ queryParams.toString() }` );
 
-		return;
+		return <></>;
```
Comment (kangzj) [inline client/my-sites/stats/stats-reditect-flow/index.tsx:80]: Any particular reason we are returning an empty fragment here?  is it possible to simply return nothing here?

File: client/my-sites/stats/controller.jsx [hunk 2]
```diff
@@ -249,7 +250,9 @@ export function redirectToDefaultModulePage( context ) {
 
 export function insights( context, next ) {
 	context.primary = (
-		<AsyncLoad require="calypso/my-sites/stats/stats-insights" placeholder={ PageLoading } />
+		<LoadStatsPage placeholder={ PageLoading }>
```
Comment (kangzj) [inline client/my-sites/stats/controller.jsx]: I'm not sure how the placeholder works - `LoadStatsPage` doesn't even have the prop declared. Can I learn more about this?

**Review by kangzj — APPROVED** [review]
It works well for me and generally looks good to me 👍 

Nitpicks:

1. Could we please use the same loading state for the purchase page? so that the transition would be smoother.

https://github.com/Automattic/wp-calypso/blob/afa39382660982914f7f356c4de8ded5c101b169/client/my-sites/stats/stats-purchase/index.tsx#L206

2. I understand some component shouldn't be applied the wrapper; however, it does look a bit more of dup code there:

<LoadStatsPage placeholder={ PageLoading }>

File: client/my-sites/stats/stats-reditect-flow/index.tsx (hunk 1 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/stats-reditect-flow/index.tsx:80]: TS was complaining about an empty return.

File: client/my-sites/stats/controller.jsx (hunk 2 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/controller.jsx]: I was experimenting with Suspense and forgot to remove this after I pulled it out.

Comment (grzegorz-cp) [reply client/my-sites/stats/controller.jsx]: It's a good point about the duplication. I'll clean that up.

**Comment by kangzj** [comment]
Update: It works well on Odyssey Stats as well. Good one! 👍

File: client/my-sites/stats/stats-reditect-flow/index.tsx (hunk 1 above)
Comment (a8ck3n) [reply client/my-sites/stats/stats-reditect-flow/index.tsx:80]: I was just reading about this yesterday and it sounds like the official recommendation is to use an explicit `null` inside components if no rendering is intended. I think this makes sense as it clarifies in the code that a return value wasn't forgotten.

I did a quick test to see if TS complained but couldn't get it to issue a warning either way (even with the empty `return`).

https://react.dev/learn/conditional-rendering#conditionally-returning-nothing-with-null

**Comment by a8ck3n** [comment]
This is a really nice change. Great to avoid the page redraw.

File: client/my-sites/stats/stats-reditect-flow/index.tsx (hunk 1 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/stats-reditect-flow/index.tsx:80]: Thanks @a8ck3n, we can address that in the maintenance project.

**Pull request merged**